src/Admin/AdminHelper.php

```diff
@@ -180,22 +179,24 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
             $objectCount = null === $value ? 0 : \count($value);
             $postCount = \count($data[$childFormBuilder->getName()]);
 
-            $fields = array_keys($fieldDescription->getAssociationAdmin()->getFormFieldDescriptions());
-
-            // for now, not sure how to do that
-            $value = [];
-            foreach ($fields as $name) {
-                $value[$name] = '';
-            }
-
             // add new elements to the subject
             while ($objectCount < $postCount) {
                 // append a new instance into the object
                 $this->addNewInstance($form->getData(), $fieldDescription);
                 ++$objectCount;
             }
 
-            $this->addNewInstance($form->getData(), $fieldDescription);
+            $newInstance = $this->addNewInstance($form->getData(), $fieldDescription);
+
+            $associationAdmin = $fieldDescription->getAssociationAdmin();
+            $associationAdmin->setSubject($newInstance);
+            $fields = array_keys($associationAdmin->getFormFieldDescriptions());
+
+            // for now, not sure how to do that
+            $value = [];
+            foreach ($fields as $name) {
+                $value[$name] = '';
```

Comment:
Where is `$value` used?

Reply:
WTF, it seems to be never used indeed. And it was already not used before my changes...

Reply:
At the beginning, the code was 
```
$this->addNewInstance($form->getData(), $fieldDescription);
        $data[$childFormBuilder->getName()][] = $value;

        $form = $admin->getFormBuilder($form->getData())->getForm();

        // bind the data
        $form->bind($data);

        $admin->setSubject($form->getData());
```

Reply:
I removed it 😅

Reply:
Damned, now I have a lower coverage 😆

Reply:
Great!
At first sight I know this will be for sure a task for another PR, but just for the records. IMO the way used to identify the form fields seem too fragile, since it's seems to be based on their order:
https://github.com/sonata-project/SonataAdminBundle/blob/9c08cef73d617fba4b36a53bc195a51a51c3040b/src/Admin/AdminHelper.php#L123-L131

Even, I don't know why `$field` is passed as reference in the `foreach`.
Whatever, later in this method, the same correlation about the form sorting is used again:
https://github.com/sonata-project/SonataAdminBundle/blob/9c08cef73d617fba4b36a53bc195a51a51c3040b/src/Admin/AdminHelper.php#L203-L209

Reply:
Indeed I would have rely on something like the $name.

This method is messy, the author knew it since he wrote
```
Note:
     *   This code is ugly, but there is no better way of doing it.
```

I agree these changes would be better in another PR.